staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -169,25 +191,37 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
 		}
 	}()
 
+	// limit to a request every half of the configured timeout with a maximum burst of one
+	// rate limited requests will receive the last request sent error (note: not the last received response)
+	limiter := rate.NewLimiter(rate.Every(timeout/2), 1)
+	// initial state is the clientErr
+	lastError := &atomicLastError{err: fmt.Errorf("etcd client connection not yet established")}
+
 	return func() error {
 		// Given that client is closed on shutdown we hold the lock for
 		// the entire period of healthcheck call to ensure that client will
 		// not be closed during healthcheck.
 		// Given that healthchecks has a 2s timeout, worst case of blocking
 		// shutdown for additional 2s seems acceptable.
-		lock.Lock()
-		defer lock.Unlock()
+		lock.RLock()
+		defer lock.RUnlock()
+
 		if clientErr != nil {
 			return clientErr
 		}
+		if limiter.Allow() == false {
+			return lastError.Load()
+		}
 		ctx, cancel := context.WithTimeout(context.Background(), timeout)
```

Comment:
is timeout guaranteed to be shorter than the timeout on the request?  Should it be?

Reply:
hmm... request isn't even plumbed to this method, so... no, the particular request triggering the call to this check could be cancelled or timeout before this completes

Reply:
I have on my TODO to understand the request path on the /healthz endpoint, but the request context is independent of the healthcheck 

https://github.com/kubernetes/kubernetes/blob/97b58ece4a2ad1a366a9cf7deea9581373931238/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go#L205-L207

but it seems this was always this way with this healthcheck

Reply:
That's true, but in this PR, one request's result can impact the results returned to other requests, so timeout/failure/cancellation is now relevant beyond just the single request

I want to make sure we don't have a weird interaction where:
* a request consumes a rate limiter token
* etcd health check gets spawned
* request cancels/terminates quickly
* timeout filter panics/interrupts the etcd health check in a way that keeps the cached result from being refreshed

Reply:
I see, I got it now, ... based on the handler code linked above,  the request lifecycle is independent of the healthcheck lifecycle ... the context passed to the etcd client probe is based on the configured healthcheckTimeout ... it seems safe

Reply:
I think with the current implementation, all in-flight `readyz` and `healthz` requests will need to contend for the lock, so in essence all healthz/readyz probes are serialized, if i am not mistaken.

https://github.com/kubernetes/kubernetes/blob/6e3d62ca1c9e1132be5cd0be6746512fdd411b60/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L172-L191

if we want a constant lookup for the probes, and we want probe requests in flight to be independent of each other then making the actual check async as shown in https://github.com/kubernetes/kubernetes/pull/104437 is a safer option.

@aojea I have a draft https://github.com/kubernetes/kubernetes/pull/112110 for reference

Reply:
Do we want to create a periodic probe?

Reply:
that is simpler than this, of course

Reply:
> I think Jordan asked the same question here , #112046 (comment)
> what timeout is the one we are taking about here? The apiserver handler has a special handler that derives the timeout form the args of the request, but I think it does not apply here
> 

@liggitt @deads2k I've asked @sttts about the handler and I was wrong, timeout on requests are respected on this handler too , I've confirmed it:

> kubectl get --raw https://1/healthz/etcd?timeout=1us
> Error from server (Timeout): the server was unable to return a response in the time allotted, but may still be processing the request

The answer should be cached though

Reply:
yeah it seems correct that this check continues even if the triggering request is cancelled/timed out.